compiler/noirc_frontend/src/tests.rs

```diff
@@ -1444,3 +1451,328 @@ fn specify_method_types_with_turbofish() {
+    // Error from the let statement annotated type
+    assert!(matches!(
+        errors[1].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+    // Error from the return type
+    assert!(matches!(
+        errors[2].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+}
+
+#[test]
+fn numeric_generic_used_in_nested_type_fail() {
+    let src = r#"
+    struct Foo<let N: u64> {
+        a: Field,
+        b: Bar<N>,
+    }
+    struct Bar<N> {
+        inner: N
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 1);
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+}
+
+#[test]
+fn numeric_generic_used_in_nested_type_pass() {
+    // The order of these structs should not be changed to make sure
+    // that we are accurately resolving all struct generics before struct fields
+    let src = r#"
+    struct NestedNumeric<let N: u64> {
+        a: Field,
+        b: InnerNumeric<N>
+    }
+    struct InnerNumeric<let N: u64> {
+        inner: [u64; N],
+    }    
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn numeric_generic_used_in_trait() {
+    let src = r#"
+    struct MyType<T> {
+        a: Field,
+        b: Field,
+        c: Field,
+        d: T,
+    }
+    
+    impl<let N: u64, T> Deserialize<N, T> for MyType<T> {
+        fn deserialize(fields: [Field; N], other: T) -> Self {
+            MyType { a: fields[0], b: fields[1], c: fields[2], d: other }
+        }
+    }
+    
+    trait Deserialize<let N: u64, T> {
+        fn deserialize(fields: [Field; N], other: T) -> Self;
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    // We want to make sure that `N` in `impl<let N: u64, T> Deserialize<N, T>` does
+    // not trigger `expected type, found numeric generic parameter N` as the trait
+    // does in fact expect a numeric generic.
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn numeric_generic_in_trait_impl_with_extra_impl_generics() {
+    let src = r#"
+    trait Default {
+        fn default() -> Self;
+    }
+
+    struct MyType<T> {
+        a: Field,
+        b: Field,
+        c: Field,
+        d: T,
+    }
+    
+    // Make sure that `T` is placed before `N` as we want to test the order is correctly maintained
+    impl<T, let N: u64> Deserialize<N> for MyType<T> where T: Default {
+        fn deserialize(fields: [Field; N]) -> Self {
+            MyType { a: fields[0], b: fields[1], c: fields[2], d: T::default() }
+        }
+    }
+    
+    trait Deserialize<let N: u64> {
+        fn deserialize(fields: [Field; N]) -> Self;
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn numeric_generic_used_in_where_clause() {
```

Comment:
Can you check that we error in the where clause as well if we pas in e.g. `T` for `Deserialize<T>` in this test

Reply:
We currently error with something other than `type used when numeric was expected" as we still allow implicit numeric generics. However, code such as the following:
```rust
    trait Deserialize<let N: u32> {
        fn deserialize(fields: [Field; N]) -> Self;
    }

    fn read<T, N>() -> T where T: Deserialize<N> {
        let mut fields: [Field; N] = [0; N];
        for i in 0..N {
            fields[i] = i as Field + 1;
        }
        T::deserialize(fields)
    }
```
will still error out with `N` not being declared rather than necessarily an error inside of `where T: Deserialize<N>`